Octo 1930 implement quantile filtering

mkolarek opened the pull request:
Implemented quantile filtering for easier filtering of entities whose KPI values exceed the value of the of the KPI at a given percentile. It is the default (and currently only) method invoked with the filter() method.

Commit:
OCTO-1930 implemented filter method, additional checks and tests needed

Commit:
OCTO-1930 implemented filter method, additional checks and tests needed

Commit:
OCTO-1930 added metadata, added tests

Commit:
OCTO-1930 added metadata, added tests

Commit:
OCTO-1930 re-enabled bayesian unit tests

Commit:
OCTO-1930 re-enabled bayesian unit tests

gbordyugov commented on expan/core/experiment.py:
```diff
@@ -141,3 +141,62 @@ def delta(self, method='fixed_horizon', **worker_args):
+        for column in df.columns:
+            if flags is None:
+                flags = df[column].apply(
+                    self._quant_apply,
```
Are we planning to use different functions here? otherwise `lambda x, y: x >= y` would suffice and there is no need to declare the static method above

mkolarek replied:
this could be simplified, true

gbordyugov commented on expan/core/experiment.py:
```diff
@@ -141,3 +141,62 @@ def delta(self, method='fixed_horizon', **worker_args):
+        else:
+            return True
+
+    def _quantile_filtering(self, df, percentile):
```
why pass the `df` parameter and not use the `data` field of the class directly?

mkolarek replied:
initially I didn't want to modify the existing data, but ultimately I decided to keep it simple... this is a leftover of the first approach, will fix

gbordyugov commented on expan/core/experiment.py:
```diff
@@ -141,3 +141,62 @@ def delta(self, method='fixed_horizon', **worker_args):
+        flags = None
+        for column in df.columns:
+            if flags is None:
+                flags = df[column].apply(
```
can we initialize `flags` with smth like `flags = [False]*len(df[column])` and drop the `if` condition altogether (since the `then` and the `else` arms of the conditional are nearly identical)?

mkolarek replied:
will look into this

Commit:
Merge remote-tracking branch 'origin/OCTO-1930-implement-quantile-filtering' into OCTO-1930-implement-quantile-filtering

# Conflicts:
#	tests/tests_core/test_experiment.py

Commit:
OCTO-1930 accidentally reverted fix, restored

gbordyugov commented on expan/core/experiment.py:
```diff
@@ -144,3 +144,62 @@ def delta(self, method='fixed_horizon', **worker_args):
+                flags = flags | df[column].apply(
+                    self._quant_apply,
+                    threshold=np.percentile(df[column], percentile)
+                )
```
or even smth like for the whole function

```
def _quantile_filtering(self, df, percentile):
  from functools import reduce

  def update_flags(flags, column):
      return flags | df[column].apply(self._quant_apply, threshold=np.percentile(df[column], percentile)

  init_flags = [False]*len(df[columns[0]])
  return reduce(update_flags, columns, init_flags)
```

gbordyugov commented on expan/core/experiment.py:
```diff
@@ -144,3 +144,62 @@ def delta(self, method='fixed_horizon', **worker_args):
+        if (len(flags[flags == True]) / len(self.data)) > 0.02:
+            warnings.warn('More than 2% of entities have been filtered out, consider adjusting the percentile value.')
+
+        self.metadata = metadata
```
We have to work out a convention about metadata: should it be a list or a dict with all possible pieces of metadata

mkolarek replied:
agreed, this is just the simplest form I could think of

gbordyugov reviewed (commented):
Just a few comments

Commit:
OCTO-1930 removed test for warning, not compatible with py27

Commit:
OCTO-1930 re-enabled bayesian tests

shansfolder commented on tests/tests_core/test_experiment.py:
```diff
@@ -241,6 +241,47 @@ def test_bayes_precision_delta_derived_kpis(self):
+        )
+        self.assertEqual(len(self.data) - len(exp.data), exp.metadata['filtered_entities_number'])
+
+    def test_quantile_filtering_2(self):
```
I would give it a more meaningful name. Also for the other quantile filtering tests.

mkolarek replied:
true, names improved

shansfolder commented on expan/core/experiment.py:
```diff
@@ -144,3 +144,62 @@ def delta(self, method='fixed_horizon', **worker_args):
+        metadata = {
+            'filtered_columns': kpis,
+            'filtered_entities_number': len(flags[flags == True])
+        }
```
We already have ```self.metadata``` when the experiment object is initialized, so I would read ```self.metadata first``` and then add these two fields to it.

Commit:
OCTO-1930
- working on self.data instead of passing as df
- improved flags init, removed if
- replaced _quant_apply with a lambda in apply()
- appended new fields to metadata instead of replacing
- improved test names

Commit:
make development requirements open ended

Commit:
simplified python package install in .travis.yml

Commit:
Merge branch 'improvement/make-dev-req-open-ended' into OCTO-1930-implement-quantile-filtering

gbordyugov commented:
Marko btw do you compare against just one percentile supposing that the outliers have always larger KPI values?

mkolarek commented:
@gbordyugov yes, I always assume that the outlier values are higher than the percentile value... this is a good point, perhaps I can add an additional parameter that indicates what kind of threshold the specified percentile should create

gbordyugov commented on expan/core/experiment.py:
```diff
@@ -144,3 +144,47 @@ def delta(self, method='fixed_horizon', **worker_args):
+        self.metadata['filtered_entities_number'] = len(flags[flags == True])
+
+        # throw warning if too many entities have been filtered out
+        if (len(flags[flags == True]) / len(self.data)) > 0.02:
```
In Python 2, an `int` divided by another `int` results in an `int`

daryadedik commented on expan/core/experiment.py:
```diff
@@ -144,3 +144,62 @@ def delta(self, method='fixed_horizon', **worker_args):
+        if value < threshold:
+            return False
+        else:
+            return True
```
i suppose, can be written simpler:

`def _quant_apply(value, threshold):
      return value > threshold`

Commit:
OCTO-1930 added threshold kind parameter for distinguishing whether a threshold is upper or lower, fixed integer division

Commit:
OCTO-1930 removed leftover print()

mkolarek commented:
@gbordyugov I've added a parameter for specifying the kind of threshold used for the filtering

shansfolder commented on expan/core/experiment.py:
```diff
@@ -144,3 +144,57 @@ def delta(self, method='fixed_horizon', **worker_args):
+        Args:
+            kpis (list): list of KPI names
+            percentile (float): percentile considered as threshold
+
```
I would document what does ```kind``` mean.
Besides, the word ```kind``` sounds kind of inaccurate. Maybe rename to ```filter_out_type```?

mkolarek replied:
good catch, forgot to document that. will rename it as well.

Commit:
OCTO-1930 renamed kind to threshold_type, documented threshold_type, refactored how threshold_type controls flow
